src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Editor/EditorStateModel.cs

```diff
@@ -0,0 +1,289 @@
+            ExportFactory<ITextBufferStateListener> textBufferListenerFactory,
+            ExportFactory<IFrameOpenCloseListener> frameEventsListenerFactory,
+            ExportFactory<ITextBufferManager> textBufferManagerFactory)
+        {
+            Requires.NotNull(threadingService, nameof(threadingService));
+            Requires.NotNull(unconfiguredProject, nameof(unconfiguredProject));
+            Requires.NotNull(serviceProvider, nameof(serviceProvider));
+            Requires.NotNull(shellHelper, nameof(shellHelper));
+            Requires.NotNull(projectFileModelWatcherFactory, nameof(projectFileModelWatcherFactory));
+            Requires.NotNull(textBufferListenerFactory, nameof(textBufferListenerFactory));
+            Requires.NotNull(frameEventsListenerFactory, nameof(frameEventsListenerFactory));
+            Requires.NotNull(textBufferManagerFactory, nameof(textBufferManagerFactory));
+
+            _threadingService = threadingService;
+            _unconfiguredProject = unconfiguredProject;
+            _serviceProvider = serviceProvider;
+            _shellHelper = shellHelper;
+            _projectFileWatcherFactory = projectFileModelWatcherFactory;
+            _textBufferListenerFactory = textBufferListenerFactory;
+            _frameEventsListenerFactory = frameEventsListenerFactory;
+            _textBufferManagerFactory = textBufferManagerFactory;
+        }
+
+        // Initialization Logic
+
+        /// <summary>
+        /// Called by anything attempting to open a project file editor window, usually by a command. This will show the window frame, creating it if
+        /// has not already been created.
+        /// </summary>
+        public async Task OpenEditorAsync()
+        {
+            await _threadingService.SwitchToUIThread();
+            lock (_lock)
+            {
+                // If the editor is already open, just show it and return
+                if (_currentState != EditorState.NoEditor)
+                {
+                    // If we're initializing, _windowFrame might be null. In that case, when the initialization code
+                    // is done, it'll take care of showing the frame.
+                    _windowFrame?.Show();
+                    return;
+                }
+                _currentState = EditorState.Initializing;
+            }
+
+            // Set up the buffer manager, which will create the temp file
+            _textBufferManager = _textBufferManagerFactory.CreateExport().Value;
+            await _textBufferManager.InitializeBufferAsync().ConfigureAwait(false);
+
+            // Future calls need to be on the UI thread, as we're creating windows.
+            await _threadingService.SwitchToUIThread();
+
+            // Open and show the editor frame.
+            _windowFrame = _shellHelper.OpenDocumentWithSpecificEditor(_serviceProvider, _textBufferManager.FilePath, XmlFactoryGuid, Guid.Empty);
```

Comment:
Wondering out loud if it's cleaner for these sorts of things to always be async, and let them switch to the UI thread under the covers. That way callers never have to worry about if they need to be on the UI thread or not.